### ISSUE TITLE
 Transfer and refactor code from message_data.reporting

### ISSUE THREAD

**Issue by khaeru** [khaeru/genno#8]
Only generalizable pieces, e.g. related to configuration, plugins, callbacks.

**Comment by khaeru** [comment]
Closed prematurely by GitHub.

**Comment by khaeru** [comment]
After #16, the callback code will be added, but not tested extensively; this is reserved for a follow-up issue.
